### PR TITLE
Version mismatch is harder to miss.

### PR DESCRIPTION
**Pull request by gerth2**

### CODE REVIEW

**Comment by gerth2** [comment]
obliterates https://github.com/PhotonVision/photonvision/issues/1072 and https://github.com/PhotonVision/photonvision/issues/1085

File: photon-lib/src/main/native/cpp/photon/PhotonCamera.cpp [hunk 1]
```diff
@@ -200,9 +200,33 @@ void PhotonCamera::VerifyVersion() {
...
+    bfw += ">>> This is neither tested nor supported.                       \n";
+    bfw += ">>> You MUST update either PhotonVision, PhotonLib, or both.    \n";
+    bfw += ">>>                                                             \n";
+    bfw += ">>> Your code will now crash. We hope your day gets better.     \n";
+    bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
+    bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
+
+    FRC_ReportError(frc::warn::Warning, "\n\n\n\n");
+    FRC_ReportError(frc::warn::Warning, bfw);
+    FRC_ReportError(frc::warn::Warning, "\n\n");
```
Comment (calcmogul) [inline photon-lib/src/main/native/cpp/photon/PhotonCamera.cpp:224]: These will show up as three separate warnings in the log, which is probably not what you want.

File: photon-lib/src/main/native/cpp/photon/PhotonCamera.cpp [hunk 2]
```diff
@@ -200,9 +200,33 @@ void PhotonCamera::VerifyVersion() {
...
+    bfw += ">>>                                                             \n";
+    bfw += ">>> You are running an incompatible version                     \n";
+    bfw += ">>> of PhotonVision on your coprocessor!                        \n";
+    bfw += ">>>                                                             \n";
+    bfw += ">>> This is neither tested nor supported.                       \n";
+    bfw += ">>> You MUST update either PhotonVision, PhotonLib, or both.    \n";
+    bfw += ">>>                                                             \n";
+    bfw += ">>> Your code will now crash. We hope your day gets better.     \n";
+    bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
+    bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
```
Comment (calcmogul) [inline photon-lib/src/main/native/cpp/photon/PhotonCamera.cpp:220]: The following is more efficient and has less visual noise. The same should work in Java.
```suggestion
    std::string_view bfw =
      ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n"
      ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n"
      ">>>    ____  _________    ____     ________  ___________    __  \n"
      ">>>   / __ \\/ ____/   |  / __ \\   /_  __/ / / /  _/ ___/   / /\n"
      ">>>  / /_/ / __/ / /| | / / / /    / / / /_/ // / \\__ \\   / / \n"
      ">>> / _, _/ /___/ ___ |/ /_/ /    / / / __  // / ___/ /  /_/    \n"
      ">>>/_/ |_/_____/_/  |_/_____/    /_/ /_/ /_/___//____/  (_)     \n"
      ">>>                                                             \n"
      ">>> You are running an incompatible version                     \n"
      ">>> of PhotonVision on your coprocessor!                        \n"
      ">>>                                                             \n"
      ">>> This is neither tested nor supported.                       \n"
      ">>> You MUST update either PhotonVision, PhotonLib, or both.    \n"
      ">>>                                                             \n"
      ">>> Your code will now crash. We hope your day gets better.     \n"
      ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n"
      ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";

```

File: photon-lib/py/photonlibpy/photonCamera.py [hunk 3]
```diff
@@ -165,11 +165,31 @@ def _versionCheck(self) -> None:
...
-                "Photon version "
-                + PHOTONVISION_VERSION
-                + " does not match coprocessor version "
-                + versionString
-                + f"! Please install photonlibpy version {PHOTONLIB_VERSION}",
-                True,
-            )
+            # Verified version mismatch
+
+            bfw = ""
```
Comment (calcmogul) [inline photon-lib/py/photonlibpy/photonCamera.py:170]: Use a multiline string instead. Those use triple quotation marks.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: photon-lib/py/photonlibpy/photonCamera.py (hunk 3 above)
Comment (srimanachanta) [reply photon-lib/py/photonlibpy/photonCamera.py:170]: Aren't those Java-17 specific? We still target java 11 in photon-lib

File: photon-lib/py/photonlibpy/photonCamera.py [hunk 4]
```diff
@@ -165,11 +165,32 @@ def _versionCheck(self) -> None:
...
+            >>> of PhotonVision on your coprocessor!
+            >>>
+            >>> This is neither tested nor supported.
+            >>> You MUST update either PhotonVision, PhotonLib, or both.
+            >>>
+            >>> Your code will now crash. We hope your day gets better.
+            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+            \n\n
+            """
```
Comment (calcmogul) [inline photon-lib/py/photonlibpy/photonCamera.py:190]: You may want to remove the leading whitespace.
```python
bfw = """
            \n\n\n
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            >>>    ____  _________    ____     ________  ___________    __
            >>>   / __ \\/ ____/   |  / __ \\   /_  __/ / / /  _/ ___/   / /
            >>>  / /_/ / __/ / /| | / / / /    / / / /_/ // / \\__ \\   / /
            >>> / _, _/ /___/ ___ |/ /_/ /    / / / __  // / ___/ /  /_/
            >>>/_/ |_/_____/_/  |_/_____/    /_/ /_/ /_/___//____/  (_)
            >>>
            >>> You are running an incompatible version
            >>> of PhotonVision on your coprocessor!
            >>>
            >>> This is neither tested nor supported.
            >>> You MUST update either PhotonVision, PhotonLib, or both.
            >>>
            >>> Your code will now crash. We hope your day gets better.
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            \n\n
            """
print(bfw)
```
```bash
[tav@myriad ~]$ python test.py

            



            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            >>>    ____  _________    ____     ________  ___________    __
            >>>   / __ \/ ____/   |  / __ \   /_  __/ / / /  _/ ___/   / /
            >>>  / /_/ / __/ / /| | / / / /    / / / /_/ // / \__ \   / /
            >>> / _, _/ /___/ ___ |/ /_/ /    / / / __  // / ___/ /  /_/
            >>>/_/ |_/_____/_/  |_/_____/    /_/ /_/ /_/___//____/  (_)
            >>>
            >>> You are running an incompatible version
            >>> of PhotonVision on your coprocessor!
            >>>
            >>> This is neither tested nor supported.
            >>> You MUST update either PhotonVision, PhotonLib, or both.
            >>>
            >>> Your code will now crash. We hope your day gets better.
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
            


            
[tav@myriad ~]$
```

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java [hunk 5]
```diff
@@ -374,13 +374,37 @@ else if (!isConnected()) {
...
+            bfw += ">>> This is neither tested nor supported.                        \n";
+            bfw += ">>> You MUST update either PhotonVision, PhotonLib, or both.     \n";
+            bfw += ">>>                                                              \n";
+            bfw += ">>> Your code will now crash. We hope your day gets better.      \n";
+            bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
+            bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
+
+            DriverStation.reportWarning("\n\n\n\n", false);
+            DriverStation.reportWarning(bfw, false);
+            DriverStation.reportWarning("\n\n", false);
```
Comment (calcmogul) [inline photon-lib/src/main/java/org/photonvision/PhotonCamera.java:399]: These should be merged together so only one warning is generated.

File: photon-lib/src/main/native/cpp/photon/PhotonCamera.cpp [hunk 6]
```diff
@@ -200,9 +200,33 @@ void PhotonCamera::VerifyVersion() {
           cameraNameOutString);
     }
   } else if (!VersionMatches(versionString)) {
-    FRC_ReportError(frc::warn::Warning,
+    std::string bfw =
```
Comment (calcmogul) [inline photon-lib/src/main/native/cpp/photon/PhotonCamera.cpp:203]:
```suggestion
    std::string_view bfw =
```

File: photon-lib/src/main/java/org/photonvision/PhotonCamera.java [hunk 7]
```diff
@@ -374,13 +374,37 @@ else if (!isConnected()) {
...
+            bfw += ">>>                                                              \n";
+            bfw += ">>> You are running an incompatible version                      \n";
+            bfw += ">>> of PhotonVision on your coprocessor!                         \n";
+            bfw += ">>>                                                              \n";
+            bfw += ">>> This is neither tested nor supported.                        \n";
+            bfw += ">>> You MUST update either PhotonVision, PhotonLib, or both.     \n";
+            bfw += ">>>                                                              \n";
+            bfw += ">>> Your code will now crash. We hope your day gets better.      \n";
+            bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
+            bfw += ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\n";
```
Comment (calcmogul) [inline photon-lib/src/main/java/org/photonvision/PhotonCamera.java:395]: Here's an option if JDK 11 compat is needed: https://godbolt.org/z/er4ME3a64

If JDK 11 compat is dropped, multiline comments are in JDK 15 and up: https://stackoverflow.com/a/50155171

Comment (mcm001) [reply photon-lib/src/main/java/org/photonvision/PhotonCamera.java:395]: We'll upgrade to Java 17 eventually so I'll go with the godbolt thing for now

**Review by calcmogul — APPROVED** [review]

**Review by mcm001 — APPROVED** [review]

**Pull request merged**